contrib/TestHarness2/test_harness/fdb.py

```diff
@@ -0,0 +1,145 @@
+                         coverage: List[Tuple[Coverage, bool]], initialized: bool) -> bool:
+    cov_dir = fdb.directory.create_or_open(tr, path)
+    if not initialized:
+        metadata_dir = fdb.directory.create_or_open(tr, metadata)
+        v = tr[metadata_dir['initialized']]
+        initialized = v.present()
+    for cov, covered in coverage:
+        if not initialized or covered:
+            tr.add(cov_dir.pack((cov.file, cov.line, cov.comment)), struct.pack('<I', 1 if covered else 0))
+    return initialized
+
+
+@fdb.transactional
+def set_initialized(tr, metadata: Tuple[str, ...]):
+    metadata_dir = fdb.directory.create_or_open(tr, metadata)
+    tr[metadata_dir['initialized']] = fdb.tuple.pack((True,))
+
+
+def write_coverage(cluster_file: str | None, cov_path: Tuple[str, ...], metadata: Tuple[str, ...],
+                   coverage: OrderedDict[Coverage, bool]):
+    db = open_db(cluster_file)
+    assert config.joshua_dir is not None
+    initialized: bool = False
+    for chunk in chunkify(coverage.items(), 100):
+        initialized = write_coverage_chunk(db, cov_path, metadata, chunk, initialized)
+    if not initialized:
+        set_initialized(db, metadata)
+
+
+@fdb.transactional
+def _read_coverage(tr, cov_path: Tuple[str, ...]) -> OrderedDict[Coverage, int]:
+    res = collections.OrderedDict()
+    cov_dir = fdb.directory.create_or_open(tr, cov_path)
+    for k, v in tr[cov_dir.range()]:
+        file, line, comment = cov_dir.unpack(k)
+        count = struct.unpack('<I', v)[0]
+        res[Coverage(file, line, comment)] = count
+    return res
+
+
+def read_coverage(cluster_file: str | None, cov_path: Tuple[str, ...]) -> OrderedDict[Coverage, int]:
+    db = open_db(cluster_file)
+    return _read_coverage(db, cov_path)
+
+
+class TestStatistics:
+    def __init__(self, runtime: int, run_count: int):
+        self.runtime: int = runtime
+        self.run_count: int = run_count
+
+
+class Statistics:
+    def __init__(self, cluster_file: str | None, joshua_dir: Tuple[str, ...]):
+        self.db = open_db(cluster_file)
+        self.stats_dir = self.open_stats_dir(self.db, joshua_dir)
+        self.stats: OrderedDict[str, TestStatistics] = self.read_stats_from_db(self.db)
+
+    @fdb.transactional
+    def open_stats_dir(self, tr, app_dir: Tuple[str]):
+        stats_dir = app_dir + ('runtime_stats',)
+        return fdb.directory.create_or_open(tr, stats_dir)
+
+    @fdb.transactional
+    def read_stats_from_db(self, tr) -> OrderedDict[str, TestStatistics]:
+        result = collections.OrderedDict()
+        for k, v in tr[self.stats_dir.range()]:
+            test_name = self.stats_dir.unpack(k)[0]
+            runtime, run_count = struct.unpack('<II', v)
+            result[test_name] = TestStatistics(runtime, run_count)
+        return result
+
+    @fdb.transactional
+    def _write_runtime(self, tr, test_name: str, time: int) -> None:
```

Comment:
Yes, I thought about this. It seems writing an additional variable and reading it would add more load on FDB than it is worth. We don't need these numbers to be super precise.